### PR TITLE
HPCC-13775 Use __sync_lock_release in spin blocks

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please can you take a close look at this.

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -473,7 +470,7 @@ class jlib_decl NonReentrantSpinLock
     { 
         ThreadId self = GetCurrentThreadId(); 
         assertex(self!=owner.tid); // check for reentrancy
-        while (!atomic_cas(&value,1,0))
+        while (!atomic_acquire(&value))
```
Comment (mckellyln) [inline system/jlib/jmutex.hpp]: Did you want unlikely() here as well ?

**Comment by mckellyln** [comment]
__sync_lock_test_and_set(&v->counter, 1)
 __sync_lock_release(&v->counter)

while correct impose a large performance penalty in benchmark tests on x86_64 for light and heavy contention.  I do not believe they are required for x86_64 ?

**Comment by mckellyln** [comment]
lock type: (21) spin-lock with back-off (atomic_t) Hash: 0               <- read on var with back off at 10
lock type: (22) spin-lock no back-off (atomic_t) Hash: 0                 <- original HPCC
lock type: (23) spin-lock aquire/release no back-off (atomic_t) Hash: 0  <- proposed acquire/release

-t  1 -l 21   2.99s user   0.00s system  99% cpu   2.995 total
-t  1 -l 22   2.97s user   0.00s system  99% cpu   2.973 total
-t  1 -l 23   6.80s user   0.00s system  99% cpu   6.812 total

-t  2 -l 21  11.03s user   3.82s system 197% cpu   7.517 total
-t  2 -l 22  14.97s user   6.29s system 198% cpu  10.694 total
-t  2 -l 23  28.53s user  11.48s system 199% cpu  20.103 total

-t  4 -l 21  27.48s user  14.58s system 352% cpu  11.930 total
-t  4 -l 22  39.32s user  23.35s system 376% cpu  16.665 total
-t  4 -l 23  63.79s user  34.13s system 326% cpu  29.945 total

-t  8 -l 21  53.60s user  32.14s system 619% cpu  13.833 total
-t  8 -l 22 179.43s user  58.66s system 695% cpu  34.245 total
-t  8 -l 23 326.95s user 107.39s system 697% cpu 1:02.28 total

-t 16 -l 21  57.24s user  49.52s system 716% cpu  14.908 total
-t 16 -l 22 164.93s user  92.59s system 767% cpu  33.557 total
-t 16 -l 23 188.67s user 158.24s system 744% cpu  46.625 total

-t 24 -l 21  65.18s user  56.48s system 740% cpu  16.428 total
-t 24 -l 22  77.46s user  96.39s system 772% cpu  22.494 total
-t 24 -l 23 156.97s user 181.74s system 777% cpu  43.581 total

-t 32 -l 21  65.89s user  57.97s system 757% cpu  16.352 total
-t 32 -l 22  89.67s user  97.41s system 764% cpu  24.456 total
-t 32 -l 23 111.39s user 169.35s system 765% cpu  36.651 total

File: system/jlib/jmutex.hpp (hunk 1 above)
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: Yes, I spotted that, but must have failed to push it.

**Comment by mckellyln** [comment]
ok, same as above but with back-off added to lock for method 23, as in:

    while (unlikely(!atomic_acquire(my_lock)))
    {
        while (unlikely(atomic_read(my_lock)))
        {
            if (i++ == maxSpins)
            {
                i = 0;
                thread_yield();
            }
        }
    }

Its a big improvement over original above but still not as good as method 21 (read on var with back-off set to 10)

-t  1 -l 21   2.95s user   0.00s system  99% cpu   2.957 total
-t  1 -l 22   3.02s user   0.00s system  99% cpu   3.025 total
-t  1 -l 23   6.81s user   0.00s system  99% cpu   6.815 total

-t  2 -l 21   7.74s user   2.16s system 176% cpu   5.620 total
-t  2 -l 22  14.60s user   6.15s system 197% cpu  10.499 total
-t  2 -l 23  15.67s user   5.77s system 193% cpu  11.081 total

-t  4 -l 21  30.41s user  15.95s system 370% cpu  12.510 total
-t  4 -l 22  33.48s user  20.96s system 374% cpu  14.524 total
-t  4 -l 23  51.85s user  32.18s system 363% cpu  23.136 total

-t  8 -l 21  61.54s user  34.45s system 650% cpu  14.751 total
-t  8 -l 22 169.50s user  56.49s system 697% cpu  32.411 total
-t  8 -l 23  86.71s user  62.61s system 632% cpu  23.608 total

-t 16 -l 21  61.44s user  51.63s system 708% cpu  15.956 total
-t 16 -l 22 111.38s user  92.26s system 769% cpu  26.449 total
-t 16 -l 23  93.40s user 104.89s system 758% cpu  26.144 total

-t 24 -l 21  61.32s user  56.56s system 728% cpu  16.183 total
-t 24 -l 22  92.76s user  99.84s system 778% cpu  24.728 total
-t 24 -l 23  89.31s user 110.95s system 773% cpu  25.887 total

-t 32 -l 21  60.47s user  59.38s system 748% cpu  16.006 total
-t 32 -l 22  94.02s user  90.50s system 763% cpu  24.176 total
-t 32 -l 23  97.77s user 124.64s system 779% cpu  28.538 total

**Comment by ghalliday** [comment]
As far as I an see from comparing the assembler, the code for the spin lock is almost the same.  The difference is that for some reason the compiler seems to think it needs an extra register.  And doesn't generate the same quality of code.

I will look at making it conditional on the architecture (I almost did anyway).

**Comment by mckellyln** [comment]
lock w/acquire:
.L3:
    call    sched_yield
.L2:
    movl    %ebp, %eax
    xchgl   (%rbx), %eax
    testl   %eax, %eax
    jne .L3

unlock w/release:
    mfence
    movl    $0, (%rdi)

lock with original cas:
.L3:
    call    sched_yield
.L2:
    movl    $1, %edx
    movl    %ebp, %eax
    lock cmpxchgl   %edx, (%rbx)
    jne .L3

unlock original:
    movl    $0, (%rdi)
    ret

**Comment by ghalliday** [comment]
Which version of gcc is that?  I have no mfence in the generated code I am looking at.  It could be context dependent I guess.

**Comment by mckellyln** [comment]
I am using 4.6.4

**Comment by ghalliday** [comment]
I was using 4.8.2 and it doesn't use a fence...  I'll still change since the assembly was better for my example.

**Comment by ghalliday** [comment]
@mckellyln please take a look at the modified pull request.  I suggest you release you improved back-off once this is merged.

File: system/jlib/jatomic.hpp [hunk 2]
```diff
@@ -148,6 +151,16 @@ static __inline__ bool atomic_cas_ptr(void **v,void *newvalue, void *expectedval
 
 #define compiler_memory_barrier() asm volatile("": : :"memory")
 
+static __inline bool atomic_acquire(atomic_t *v)
+{
```
Comment (mckellyln) [inline system/jlib/jatomic.hpp]: __inline or __inline__ or does it not matter ?

Comment (ghalliday) [inline system/jlib/jatomic.hpp]: It should be consistent - I'll change

**Comment by mckellyln** [comment]
Looks good to merge.

**Pull request merged**